### PR TITLE
Add Three point error handling to borrowck

### PR DESCRIPTION
**Pull request by spastorino**
This is for https://github.com/rust-lang/rust/issues/45988

### CODE REVIEW

File: src/librustc_mir/borrow_check/mod.rs [hunk 1]
```diff
@@ -2693,7 +2678,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
 }
 
 #[derive(Copy, Clone, PartialEq, Eq, Debug)]
-struct Context {
+pub struct Context {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: this probably does not need to be `pub`...? in particular, submodules can still access it when it is priv

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        &self,
+        context: Context,
+        borrow: &BorrowData<'_>,
+        err: &mut DiagnosticBuilder<'_>,
+    ) {
+        if let Some(regioncx) = &self.nonlexical_regioncx {
+            if let Some(cause) = regioncx.why_region_contains_point(borrow.region, context.loc) {
+                let mir = self.mir;
+
+                cause.label_diagnostic(mir, err);
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: maybe we should move this call to only execute if `cause.root_cause()` returns something other than `LiveVar` or `DropVar`... that would avoid dump extra causal information when we can give a more user-friendly version. But really I guess we just want to make the "cause-gathering" stuff enabled or lazy, so never mind.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Let's fix the `pub` thing -- as for the UI tests,  if you modify the code in `region_infer/mod.rs` and find the place where it enables tracing only if `nll_dump_cause` is true -- just make it enable tracing unconditionally, then we don't need to pass `-Znll-dump-cause`

File: src/librustc_mir/borrow_check/mod.rs (hunk 1 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/mod.rs]:
```
error[E0446]: private type `borrow_check::Context` in public interface
  --> src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs:21:5
   |
21 | /     pub fn explain_why_borrow_contains_point(
22 | |         &self,
23 | |         context: Context,
24 | |         borrow: &BorrowData<'_>,
...  |
78 | |         }
79 | |     }
   | |_____^ can't leak private type

error: aborting due to previous error
```

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs (hunk 2 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: :+1:, so should just leave as is ?

**Comment by spastorino** [comment]
> Let's fix the pub thing -- as for the UI tests, if you modify the code in region_infer/mod.rs and find the 
> place where it enables tracing only if nll_dump_cause is true -- just make it enable tracing 
> unconditionally, then we don't need to pass -Znll-dump-cause

I don't understand what you mean here. Enabling this unconditionally won't in some way rollback your work? in that case isn't it better to remove the commit that adds is from code

**Comment by nikomatsakis** [comment]
> I don't understand what you mean here. Enabling this unconditionally won't in some way rollback your work? in that case isn't it better to remove the commit that adds is from code

Sure. Just make a single commit and I'll rebase it later and fixup.

File: src/librustc_mir/borrow_check/mod.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/mod.rs]: change that `pub fn explain` to `pub(super) fn explain`

Comment (spastorino) [reply src/librustc_mir/borrow_check/mod.rs]: :+1:

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**